File: frontend/src/api/types/bookmark.ts
export type Bookmark = {
  id: number
  user_id: number
  bookmark_type: 'artist' | 'album'
  musicbrainz_id: string
  title: string
  extra_data?: string | null
  categories: BookmarkCategory[]
  created_at: string
  updated_at: string
}

export type BookmarkCategory = {
  id: number
  name: string
  is_default: boolean
  user_id?: number | null
  created_at: string
  updated_at: string
}

export type BookmarkCreate = {
  bookmark_type: 'artist' | 'album'
  musicbrainz_id: string
  title: string
  extra_data?: string | null
  category_ids?: number[]
}

export type BookmarkCategoryCreate = {
  name: string
}

File: backend/app/schemas/bookmark.py
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel


class BookmarkCategoryBase(BaseModel):
    name: str


class BookmarkCategoryCreate(BookmarkCategoryBase):
    pass


class BookmarkCategoryRead(BookmarkCategoryBase):
    id: int
    is_default: bool
    user_id: Optional[int] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class BookmarkBase(BaseModel):
    bookmark_type: str
    musicbrainz_id: str
    title: str
    extra_data: Optional[str] = None


class BookmarkCreate(BookmarkBase):
    category_ids: Optional[List[int]] = None


class BookmarkUpdate(BaseModel):
    title: Optional[str] = None
    extra_data: Optional[str] = None
    category_ids: Optional[List[int]] = None


class BookmarkRead(BookmarkBase):
    id: int
    user_id: int
    categories: List[BookmarkCategoryRead] = []
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
